### ISSUE TITLE
Where in the proposed folder structure do I put shared files like filters and services?

### ISSUE THREAD

**Issue by burzum** [ngbp/ngbp#345]
Where in the proposed folder structure do I put shared files like filters and services?

It looks to me like the project is set up to have one folder or module per controller. But where do I put services that are shared between them? Filters? Directives?

**Comment by tgallant** [comment]
you can put directives and any other shared code in `src/common`. Here is an example from the angular-app repo https://github.com/angular-app/angular-app/tree/master/client/src/common

**Comment by joshdmiller** [comment]
I agree with @tgallant, but I want to make one minor clarification. ngbp is set up to have one file per module, which often works out to one module per folder. That applies to one controller per folder, but also to one filter, one service, etc. So it's more broad is my point.

I'll go ahead and close this, but please follow up if you have more questions!

**Comment by jasny** [comment]
It's still unclear to me where I can put application specific services, filters and directives. Neither the `app` or the `common` directory seem to apply.

### src/app

> Apart from `app.js` and its accompanying tests (discussed below), this directory is filled with subdirectories corresponding to high-level sections of the application, often corresponding to top-level routes.

This is not the case, these do not correspond to a high-level section of the application, let alone to routes.

### src/common

> Every component contained here should be drag-and-drop reusable in any other  project; they should depend on no other components that aren't similarly drag-and-drop reusable.

Definitely not true in my case, these items are very much application specific and don't make sense in other applications.

_[Also asked on StackOverflow](http://stackoverflow.com/questions/28453388/where-should-i-place-application-specific-directives-with-ngboilerplate)_

**Comment by saibotsivad** [comment]
I'll tell you what we do where I work, maybe it will be helpful. Note that I'm not saying this is The Right Way, but it seems to work pretty well and be pretty manageable from a maintenance and upgradeability perspective.

Although we've modified the ngbp setup a bit, the following details work with the vanilla ngbp.

### our application

Our application is reasonably sizable:
- many different screens/pages/routes
- each screen has some level of shared state (e.g. date fields, text filters)
- each screen has multiple interactions (e.g., adjust the dates, change the filters)
- some state is only shared with some screens (e.g., maybe one date is shared between A and B, and another is shared between B and C)
- easily over 100 each of directives and factories, and probably around 20-50 controllers
- lots of old gross code (mixing factories and directives, not testable, etc.)
- multiple apps (each a new clone of ngbp) that share common code from `src/common` (we basically do `ln -s` on the folder)

Anyway, it's a sizable app, and we keep at least one full time developer on it at all times, just upgrading and fixing bugs.

### folder structure

We have it set up with `src/app` and `src/common` set up as the wording implies. The `src/common` has stuff that's "drag-and-drop reusable", while the `src/app` has application-specific code.

Here's what that actually looks like (with some actual examples):

```
src/common/autocomplete
    autocomplete.js # directives and factories specific to an input text autocomplete thing
    autocomplete.less # styles, namespaced to the autocomplete directive
src/common/jquery-dropdown
    jquery-dropdown.js # a third-party plugin we use, but modify heavily and wrapped to angularjs
    jquery-dropdown.less # the styles to touch up our mods
src/common/fallback-image
    fallback-image.js # a directive that sets a fallback image if `ng-src` doesn't load
src/common/loading-bar
    loading-bar.js # directive of a loading bar with extra options
    loading-bar.less # some style for the loading bar
src/common/days-between-dates
    days-between-dates.js # the factory
    days-between-dates.spec.js # actual tests
# etc.
```

Some observations:
- Some of these are really small files. For example, the `days-between-dates` [is only 11 lines](https://gist.github.com/saibotsivad/781b1215b145032c5536).
- You'll notice our test coverage is pretty poor. That's the fault of big old software, and it makes me sad.
- The things are generally drag-and-drop, but obviously there's some small choices (e.g. the LESS files) that make it not entirely generic.
- Although the naming scheme is redundant (having `my-module/my-module.js` used to bother me), what it allows is easier debugging in Chrome. (With the developer console open, you can hit command/control+o and type in the file name to open it. Easier than having 100 files named `index.js` or similar.)
- Not everything could be immediately plucked out and set up as a third-party module, but it can be plucked out and used between our differently-styled applications

Inside the `src/app` we've got the application-specific code. This includes "screens" (we use the wonderful [ui-router](https://github.com/angular-ui/ui-router) for states) and services that are explicitly for this application.

The pages look like:

```
src/app/splash
    /login # handle login requests (not behind password)
        login.html
        login.less
        login.js # this has the routing setup and a controller
    /password-reset # handle password resets and the email links go here
        password-reset.html
        password-reset-request.html
        password-reset.js # routing setup and controller
src/app/overlap # an example page showing overlapping data
    overlap.js # the routing setup and controller
    overlap.html
    overlap.less
    overlap-model.js # factories that handle $http.get() so that isn't in the controller
    /company-address # a directive that is only used within `overlap`
        company-address.js
        company-address.spec.js #tests
    /refinements # a child route of the parent `overlap` path
        refinements.js # routing and controller
        refinements.html
        refinement.less
src/app/audit
    audit.js # a service that we use in some of our screens to log user interaction
src/app/session
    session.js # a service we use to store things in-session and sync with the server
    session.spec.js # tests for the service
src/app
    app.js # the startup code
```

Observations:
- All the application logic sits in `src/app`, meaning routes and controllers for pages
- Directives and factories that are specific to a route/page go inside that folder (easier to find)
- We put all our $http.get() logic inside factories, and those then go in `NAME-model.js`, we've found this easy to understand and maintain

### other thoughts

In general we've found this layout to work well with our team, and to be maintainable. After having been through one big Angular upgrade (1.0 to 1.3) and then another big Bootstrap upgrade (2.1 to 3.3), I feel like this is a pretty solid setup.

Of course, YMMV, and this is only one opinion, based on only a couple applications. You may find that your application structure has much more code shared between pages. We found that ~70% of our directives and service were bound to a single page, and ~20% were abstractable enough to strip them down and put them in `src/common` to be shared between applications.

What you seem to be wondering about is the (for us, anyway) remaining 10%. For us, we found that they could be put into folders and such, like the `src/app/session` code. (That's a couple factories that sync'd session and server state. It was app-wide, not for a page.)

### in conclusion

I don't think that answers your question directly, but maybe it will inform your design decisions.

If you have more questions, feel free to ask.
